core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1602,8 +1602,6 @@ protected void updateFetchOffsetMetadata(Optional<LogOffsetMetadata> fetchOffset
     protected Optional<LogOffsetMetadata> fetchOffsetMetadata() {
         lock.readLock().lock();
         try {
-            if (findNextFetchOffset.get())
```

Comment:
>  Since we update fetchOffsetMetadata everytime we change endOffset, I don't think that we should have dependency on findNextFetchOffset while getting the value of fetchOffsetMetadata

Hmm, the issue is that nextFetchOffset doesn't return endOffset if findNextFetchOffset is true. Currently we only reset fetchOffsetMetadata when updating the endOffset. It's possible that findNextFetchOffset stays on for multiple fetches without changing endOffset. In that case, we will set fetchOffsetMetadata for the first fetch and keep reusing it for subsequent fetches, which will be incorrect.

Reply:
Hi @junrao, so shall we do this, that when the call goes to `fetchOffsetMetadata()`, we check if `findNextFetchOffset` is true or not, in case it is true, we do a call to `nextFetchOffset()` which will correctly update the endOffset if it needs to be updated or not. Finally, we just return `fetchOffsetMetadata`. Do you think it will work?

Reply:
Now that I think more about it, IIUC, considering the common case when all fetched data is acquirable -  
1. acknowledgements/acquisition lock timeout/ release of records on session close are the only places where we set `findNextFetchOffset` to true 
2. In all the 3 scenarios mentioned above, if there is a change to the `endOffset`, we update the `endOffset` (thereby `fetchOffsetMetadata` is also updated automatically with our changes)
Hence, I feel that the findNextFetchOffset shouldn't be considered when dealing with the common case. 
In the not common cases, when Log Start Offset is later than the fetch offset and we need to archive records, then we set `findNextFetchOffset` to True. But we have done the minBytes implementation only for the common cases right now, hence i feel the current change is correct. Please correct me if I am wrong.
cc - @apoorvmittal10

Reply:
Yes, I agree findNextFetchOffset=true is the uncommon case. It might be useful to at least have some kind of consistent behavior for the uncommon case. Since the minByte estimation will be off anyway in this case, we could choose to consistent satisfy the request immediately or wait for the timeout. With the logic in this PR, because fetchOffsetMetadata can be outdated in this uncommon case, it's not clear when the request will be satisfied.

Reply:
so, should I do what I suggested https://github.com/apache/kafka/pull/17739#discussion_r1838763754 here, that when the call goes to fetchOffsetMetadata(), we check if findNextFetchOffset is true or not, in case it is true, we do a call to nextFetchOffset() which will correctly update the endOffset if it needs to be updated or not. Finally, we just return fetchOffsetMetadata.

Reply:
Hmm, my understanding is that endOffset always points to the endOffset of all cached batches in sharePartition. nextFetchOffset() doesn't update endOffset. It only turns findNextFetchOffset off when the next fetch offset reaches endOffset. It could take multiple nextFetchOffset() calls before findNextFetchOffset is reset to off.

Reply:
For consistency, we can do one change - 
```
protected Optional<LogOffsetMetadata> fetchOffsetMetadata() {
        lock.readLock().lock();
        try {
            if (fetchOffsetMetadata.isPresent() && findNextFetchOffset.get()) {
                return Optional.empty();
            }
            return fetchOffsetMetadata;
        } finally {
            lock.readLock().unlock();
        }
    }
```
by this logic, we'll be able to avoid any reads where `fetchOffsetMetadata` is old and we don't want to use it. Wdyt?

Reply:
@adixitconfluent Here is my understanding: We want to keep `fetchOffsetMetadata` in SharePartition so when there is no change to endOffset then existing `fetchOffsetMetadata` can be used to calculate min bytes, correc?

For this, there could be couple of scenarios
- end offset is just increasing monotonically (usually), hence next fetch offset shall point to new latest.
- there is reset of some messages and we need to determine next fetch offset

For first I see you reset the `fetchOffsetMetadata` whenever you change the `endOffset`. For latter you use `findNextFetchOffset` always use Optional.empty to be sent. So the existing code seems fine to me.

The current code solves mostly all but there could be an edge case where only single message is released from between and the batch that holds that released message might not be able to alone satisfy min bytes criteria, is that something we are solving?

Can you confirm on my above understanding, if everything is true then the approach I suggest to have local copy of intermediate offset and metadata in delayed fetch as well. And change `updateFetchOffsetMetadata` in SharePartition to `maybeUpdateFetchOffsetMetadata(Optional<LogOffsetMetadata> fetchOffsetMetadata, long fetchOffset). The method caches the `Optional<LogOffsetMetadata> fetchOffsetMetadata` only if `fetchOffset` matches `endOffset`.

Reply:
@apoorvmittal10:
>We want to keep fetchOffsetMetadata in SharePartition so when there is no change to endOffset then existing fetchOffsetMetadata can be used to calculate min bytes, correc?

Yes

>The current code solves mostly all but there could be an edge case where only single message is released from between and the batch that holds that released message might not be able to alone satisfy min bytes criteria, is that something we are solving?

We are only solving common case where all fetched data is acquirable. So, this case is something we're not solving right now

Reply:
Thanks @adixitconfluent for confirming. The simplest way I can think of to handle common and uncommon case:

- We shall maintain `fetchOffsetMetadata` only corresponding to the end offset and not for intermediate states, as that will are most common case i.e. fetch from last.
- The endOffset shall always point to the first offfset of any log batch hence if we fetch `fetchOffsetMetadata` for end offset then it should match with `fetchOffsetMetadata.messageOffset`.

- DelayedShareFetch arrives with fetchOffset which could match the partitions endOffset or anywhere earlier, delayed share fetch doesn't know. Hence call `fetchOffsetMetadata()` for share partition which if provides the `fetchOffsetMetadata` then check:
        *  if the messageOffset matches with fetchOffset. If yes then we already have latest copy.
        * If empty, then we anyways need to readFromLog and update `fetchOffsetMetadata`.
        * If it is ahead of fetchOffset then we are fetching for released records. Hence we `readFromLog` and keep the `fetchOffsetMetadata` for the partition in DelayedShareFetch itself, which should be re-used. I am suggesting to keep it locally itself because if timeout occurs for request then anyways we will complete the request with whatever we can fetch and there is guaranteed that some messages will be acquired. Hence this intermediate `fetchOffsetMetadata` is only relevant for single session of DelayedShareFetch request. 
     
- Change `updateFetchOffsetMetadata(Optional<LogOffsetMetadata> fetchOffsetMetadata)` in SharePartition to `boolean maybeUpdateFetchOffsetMetadata(Optional<LogOffsetMetadata> fetchOffsetMetadata, long fetchOffset)` and only keep the `fetchOffsetMetadata` for the endOffset. Check if the fetchOffset matches the endOffset.
- `fetchOffsetMetadata()` should just return stored `fetchOffsetMetadata` without any check.

@junrao wdyt? Is there gap in my overall understanding?

Reply:
@apoorvmittal10: Overall, your approach could work.
1. It seems that fetchOffsetMetadata() needs to take in nextFetchOffset too. This way, if the offset matches, we can avoid the readFromLog call.
2. We want to be a bit careful with the offset matching. In the common case, the nextFetchOffset is always aligned on batch boundary so it will match fetchOffsetMetadata.messageOffset. However, a user could initialize an arbitrary offset in the middle of a batch. In this case, nextFetchOffset won't match fetchOffsetMetadata.messageOffset since the latter is at the batch boundary. So, we need to understand the impact of that.

Reply:
>It seems that fetchOffsetMetadata() needs to take in nextFetchOffset too. This way, if the offset matches, we can avoid the readFromLog call.

Sure we can have it but I was just thinking that if a share fetch request reads the nextFetchOffset as latest (endOffset) and waits in tryComplete as there are not enought bytes. But before next tryComplete iteration an offset for respective share-partition is released then nextFetchOffset will be true. In the next iteration of tryComplete, fetchOffsetMetadata will be read as Optional.empty from fetchOffsetMetadata() call, which will be incorrect. Hence I was thinking to not have this check rather work with fetchOffsetMetadata's messageOffset.

>However, a user could initialize an arbitrary offset in the middle of a batch.

Hmmm, I am just thinking if it could ever occur. As we always align on batch boundaries i.e. either fetch with max fetch records or without, we acquire through full produce batch. But in corner case, if you mean that user sets the offset to a specific offset or resets offset by duration then it could be arbitrary i.e. in between the batch itself. In that case we should treat that with local stored fetchOffsetAndMetadata copy in DelayedShareFetch.

Reply:
I was thinking of the following. Each tryComplete always calls sharePartition.nextFetchOffset() first. When we get the nextFetchOffset, we call fetchOffsetMetadata(nextFetchOffset) to see if the offsetMetadata for nextFetchOffset is available. If not, we call readFromLog to get offsetMetadata and then call updateFetchOffsetMetadata(fetchOffsetMetadata, nextFetchOffset). Inside this method, we cache both inputs. In fetchOffsetMetadata(nextFetchOffset), we compare the input with the cached nextFetchOffset to determine if the offsetMetadata is available. This covers the corner case where the fetch offset is not on the batch boundary too.

Reply:
>but I wanted to confirm whether it's necessary

Yes it's an optimization over current implemetation where we might do unnecessary processing because of released records.

Reply:
@AndrewJSchofield, in theory, AdminClient.alterShareGroupOffsets() can initialize to an arbitrary offset right?

Reply:
Yes, indeed. Great point.

Reply:
hi everyone, thanks for your inputs. @junrao , I think it is gonna be a considerable change for this PR, hence I have created JIRA https://issues.apache.org/jira/browse/KAFKA-18022 to track this issue. I would like to address this in a new PR, if it is fine to you. Meanwhile I am reverting the code change for this line and updating the PR description to reflect the same.